Make Perspective Broker's NewCredTests use IOPump

markrwilliams opened the pull request:
http://twistedmatrix.com/trac/ticket/8792

markrwilliams committed:
Draft of IOPump-based NewCredTests

markrwilliams committed:
Remove dead debugging code in test_pb, declare stub methods as no-ops, and fix a typo.

markrwilliams committed:
Remove meaningless comment and delete unused helper NewCredTests._disconnect

markrwilliams committed:
Add topfile

markrwilliams committed:
flake8

markrwilliams committed:
Merge branch 'trunk' into 8792-pb-newcredtests-iopump

markrwilliams committed:
Merge branch 'trunk' into 8792-pb-newcredtests-iopump

markrwilliams committed:
Play a branch coverage game

glyph committed:
Merge branch 'trunk' into 8792-pb-newcredtests-iopump

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -170,6 +169,52 @@ def disconnectClientFactory():
 
 
+@implementer(interfaces.IConnector)
+class ReconnectingConnector(object):
```
Please file a ticket for putting this into `proto_helpers`.  Given that this is a fake of a core Twisted interface, it shouldn't be squirreled away inside PB's tests.

glyph replied:
Actually, there's already https://github.com/twisted/twisted/blob/c02c19441f6bd9d2431ac019ba36ac6abe33cfea/src/twisted/test/proto_helpers.py#L343.  Would it be too much trouble to refactor this to use that existing fake?

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -170,6 +169,52 @@ def disconnectClientFactory():
+    """
+
+    def __init__(self, reconnectCallback, *args, **kwargs):
+        self.reconnectCallback = reconnectCallback
```
As per https://glyph.twistedmatrix.com/2015/05/separate-your-fakes-and-your-inspectors.html we should not expose any public attributes on the test versions of stuff that are not present on the "real" versions of stuff.  These should all be `_`-prefixed.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -170,6 +169,52 @@ def disconnectClientFactory():
+    An L{IConnector} implementation that calls the specified callback
+    when its C{connect} method is called
+
+    @param reconnectCallback: the callback to call when C{connect} is
```
These should be `@ivar`; "param" is invalid markup for a class.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -170,6 +169,52 @@ def disconnectClientFactory():
+        called
+    @type reconnectCallback: A callable
+
+    @param *args: Position arguments to pass C{reconnectCallback}
```
As above, including `*args` is invalid epytext and this should be just `@param args`.

Why isn't pydoctor catching this, I wonder?  Hmm.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -170,6 +169,52 @@ def disconnectClientFactory():
+    @type reconnectCallback: A callable
+
+    @param *args: Position arguments to pass C{reconnectCallback}
+    @param **kwargs: Keyword arguments to pass to C{reconnectCallback}
```
As above, `kwargs`, no stars.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1434,19 +1511,34 @@ def test_loginLogout(self):
+        self.pump.flush()
+        # Now allow the client to disconnect.
+        loginCompleted.callback(None)
         return d
```
Synchronous tests should no longer return `Deferred`s.  This is what tells Twisted to run the main loop.

markrwilliams replied:
Switching over to `SynchronousTestCase` exposes a bug in PB's error handling.  The [gory details](http://twistedmatrix.com/trac/ticket/8792#comment:5) are in Trac, but it suffices to say that fixing PB's error handling deserves [its own ticket](http://twistedmatrix.com/trac/ticket/8803).

In the meantime, is it OK to continue to use an asynchronous test case that returns `Deferred`s?  Another alternative would be to write a local helper that acts like `t.t_async.TestCase.assertFailure` and use it instead of `failureResultOf`, but it seems like a bad idea to heap a temporary work around into this PR.

**Addendum**: Using `t.t._async.TestCase` with an `IOPump` _is_ silly, but it does result in a version of these tests that pass under PyPy.  So it has some merit!

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1472,23 +1564,31 @@ def logout(self):
+        gc.collect()
+        # push their decref messages through
+        self.pump.flush()
         return d
```
As above.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1518,10 +1617,10 @@ def cbAvatarIds(x):
+        self.establishClientAndServer()
+        self.pump.flush()
+
         return d
```
As above.

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1561,19 +1659,16 @@ def test_anonymousLogin(self):
-        self.addCleanup(connector.disconnect)
+        self.establishClientAndServer()
+        self.pump.flush()
         return d
```
"

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1584,18 +1679,17 @@ def test_anonymousLoginNotPermitted(self):
+        self.establishClientAndServer()
+        self.pump.flush()
+
         return d
```
"

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1607,19 +1701,17 @@ def test_anonymousLoginWithMultipleCheckers(self):
 
-        connector = reactor.connectTCP('127.0.0.1', self.portno, factory)
-        self.addCleanup(connector.disconnect)
         return d
```
"

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1641,10 +1732,9 @@ def cbLogin(perspective):
 
-        connector = reactor.connectTCP('127.0.0.1', self.portno, factory)
-        self.addCleanup(connector.disconnect)
         return d
```
"

glyph commented on src/twisted/spread/test/test_pb.py:
```diff
@@ -1669,10 +1758,9 @@ def cbView(viewpoint):
 
-        connector = reactor.connectTCP("127.0.0.1", self.portno, factory)
-        self.addCleanup(connector.disconnect)
         return d
```
"

markrwilliams committed:
twisted.spread.test_pb.ReconnectingConnector becomes _ReconnectingFakeConnector

It now inherits from twisted.test.proto_helpers._FakeConnector, which
provides most of the IConnector interface.  Thanks for the suggestion
during review, @glyph!

markrwilliams committed:
twisted.spread.test_pb.NewCredTests becomes a SynchronousTestCase

Thanks for the review, @glyph!

markrwilliams committed:
Clean up docstrings on _ReconnectingFakeConnector

markrwilliams committed:
Revert "twisted.spread.test_pb.NewCredTests becomes a SynchronousTestCase"

This reverts commit 5c19886c28262959beafd07758e3c7ab41cb7bf3.

This is reverted because SynchronousTestCase exposes bugs in PB's
error handling that will be handled under a different ticket.

markrwilliams committed:
Merge branch 'trunk' into 8792-pb-newcredtests-iopump

markrwilliams committed:
Split out _ReconnectingFakeConnector notifications into _ReconnectingFakeConnectorState

glyph committed:
Merge branch 'trunk' into 8792-pb-newcredtests-iopump

glyph commented:
Hopefully the build gods will be kinder to us with this trunk update, and I will land forthwith.

glyph commented:
I'm going to use admin powers to override 😢  because both build failures are intermittent issues with @Lukasa's Microsoft™ HomePage™ Server™ (i.e. with fetching the URL https://python-hyper.org/h2/en/stable/objects.inv).  We should _really_ fix that issue, and check in a fallback `objects.inv`.

markrwilliams commented:
@glyph thanks!  and [oh look there's a new ticket](https://twistedmatrix.com/trac/ticket/8823) now that this one's closed, proving that the number of open tickets, like energy, is always conserved.
